src/Microsoft.IdentityModel.Tokens/Encryption/KeyWrapProvider.cs

```diff
@@ -0,0 +1,439 @@
+
+            if (_symmetricAlgorithmDecryptor == null)
+            {
+                lock (decryptorLock)
+                {
+                    if (_symmetricAlgorithmDecryptor == null)
+                        _symmetricAlgorithmDecryptor = _symmetricAlgorithm.CreateDecryptor();
+                }
+            }
+
+            byte[] block = new byte[16];
+
+            // Calculate intermediate values
+            for (var j = 5; j >= 0; j--)
+            {
+                for (var i = n; i > 0; i--)
+                {
+                    // T = ( n * j ) + i
+                    var t = (ulong)((n * j) + i);
+
+                    // B = AES-1(K, (A ^ t) | R[i] )
+
+                    // First, A = ( A ^ t )
+                    a.Xor(GetBytes(t), true);
+
+                    // Second, block = ( A | R[i] )
+                    Array.Copy(a, block, BlockSizeInBytes);
+                    Array.Copy(r, (i - 1) << 3, block, BlockSizeInBytes, BlockSizeInBytes);
+
+                    // Third, b = AES-1( block )
+                    var b = _symmetricAlgorithmDecryptor.TransformFinalBlock(block, 0, 16);
+
+                    // A = MSB(64, B)
+                    Array.Copy(b, a, BlockSizeInBytes);
+
+                    // R[i] = LSB(64, B)
+                    Array.Copy(b, BlockSizeInBytes, r, (i - 1) << 3, BlockSizeInBytes);
+                }
+            }
+
+            if (a.SequenceEqualConstantTime(iv))
+            {
+                var c = new byte[n << 3];
+
+                for (var i = 0; i < n; i++)
+                {
+                    Array.Copy(r, i << 3, c, i << 3, 8);
+                }
+
+                return c;
+            }
+            else
+            {
+                throw LogHelper.LogException<InvalidOperationException>(LogMessages.IDX10665);
+            }
+        }
+
+        private void ValidateKeySize(byte[] key, string algorithm)
+        {
+            if (SecurityAlgorithms.Aes128KW.Equals(algorithm, StringComparison.Ordinal))
+            {
+                if (key.Length != 16)
+                    throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException("key.KeySize", string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10662, SecurityAlgorithms.Aes128KW, 128, Key.KeyId, key.Length << 3)));
+
+                return;
+            }
+
+            if (SecurityAlgorithms.Aes256KW.Equals(algorithm, StringComparison.Ordinal))
+            {
+                if (key.Length != 32)
+                    throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException("key.KeySize", string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10662, SecurityAlgorithms.Aes256KW, 256, Key.KeyId, key.Length << 3)));
+
+                return;
+            }
+
+            throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(algorithm), String.Format(CultureInfo.InvariantCulture, LogMessages.IDX10652, algorithm)));
```

Comment:
We are inconsistent using both String.Format and string.Format, should pick one. string is an alias for String.